File: 240713/핀볼게임/pinball-game.js
const fs = require("fs")
const input = fs.readFileSync(0).toString().trim().split("\n")
const n = Number(input[0])
const matrix = input.slice(1).map(line => line.split(" ").map(Number))

function changeDir(dir, n) {
    if (n === 1) {
        if (dir[0] === 0 && dir[1] === -1) 
            dir = [-1, 0]; 
        else if (dir[0] === -1 && dir[1] === 0)
            dir = [0, -1];
        else if (dir[0] === 0 && dir[1] === 1) 
            dir = [1, 0]; 
        else if (dir[0] === 1 && dir[1] === 0) 
            dir = [0, 1]; 
    } else if (n === 2) {
        if (dir[0] === 0 && dir[1] === -1) 
            dir = [1, 0]; 
        else if (dir[0] === -1 && dir[1] === 0) 
            dir = [0, 1]; 
        else if (dir[0] === 0 && dir[1] === 1) 
            dir = [-1, 0]; 
        else if (dir[0] === 1 && dir[1] === 0) 
            dir = [0, -1]; 
    }
    return dir;
}

function time(x, y, dir){
    let t = 1

    while(x >= 0 && x < n && y >= 0 && y < n){
        dir = changeDir(dir, matrix[y][x])
        x += dir[0]
        y -= dir[1]
        t++
    }
    return t
}

for (let i = 0; i < n; i++){
    ans = Math.max(time(i, 0, [1, 0]), time(n - 1, 0, [0, 1]), time(i, n - 1, [-1, 0]), time(0, i, [1, 0]), ans)
}
console.log(ans)

File: 240715/핀볼게임/pinball-game.py
n = int(input())
grid = [list(map(int, input().split(" "))) for _ in range(n)]

def changePosition(x, y, direction):
    dx = [1, 0, -1, 0]
    dy = [0, -1, 0, 1]

    if grid[x][y] == 2:
        direction = 3 - direction
    elif grid[x][y] == 1:
        if direction == 0 or direction == 2:
            direction += 1
        else:
            direction -= 1
    x += dx[direction]
    y += dy[direction]
    
    return x, y, direction

def calcTime(x, y, direction):
    time = 1
    while 0 <= x < n and 0 <= y < n:
        x, y, direction = changePosition(x, y, direction)
        time += 1
    return time

time = 0
for i in range(n):
    time = max(time, calcTime(0, i, 0), calcTime(i, n - 1, 1), calcTime(n - 1, i, 2), calcTime(i, 0, 3))

print(time)